src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -94,6 +99,64 @@ upgrade_to_reference_table(PG_FUNCTION_ARGS)
 
 
 /*
+ * ReplicateAllReferenceTablesToAllNodes function finds all reference tables and
+ * replicates them to all worker nodes. It also modifies pg_dist_colocation table to
+ * update the replication factor column. This function skips a worker node if that node
+ * already has healthy placement of a particular reference table to prevent unnecessary
+ * data transfer.
+ */
+void
+ReplicateAllReferenceTablesToAllNodes()
+{
+	List *referenceTableList = ReferenceTableList();
+	ListCell *referenceTableCell = NULL;
+
+	/* we do not use pgDistNode, we only obtain a lock on it to prevent modifications */
+	Relation pgDistNode = heap_open(DistNodeRelationId(), AccessShareLock);
+
+	List *workerNodeList = WorkerNodeList();
+	int workerCount = list_length(workerNodeList);
+
+	uint32 currentColocationId = INVALID_COLOCATION_ID;
+
+	foreach(referenceTableCell, referenceTableList)
+	{
+		Oid referenceTableId = lfirst_oid(referenceTableCell);
+		List *shardIntervalList = LoadShardIntervalList(referenceTableId);
+		ShardInterval *shardInterval = (ShardInterval *) linitial(shardIntervalList);
+		uint64 shardId = shardInterval->shardId;
+
+		LockShardDistributionMetadata(shardId, ExclusiveLock);
+
+		ReplicateShardToAllWorkers(shardInterval);
+
+		/* we have this check to prevent accessing to cache multiple times */
+		if (currentColocationId == INVALID_COLOCATION_ID)
```

Comment:
I think this also makes the loop harder to read. One suggestion for this:
```C

if (list_lenght(ReferenceTableList) == 1)
return

lock pg_worker_node

firstRefTable = linitial
colocationId = ColocationId(firstRefTable);

for each ref table
     log NOTICE
     lock table
     Replicate

UpdateColocationGroup()
```

what do you think?